tests/test_guess_objects.py

```diff
@@ -0,0 +1,93 @@
+import unittest
+import math
+import os
+from requre.objects import ObjectStorage
+from requre.utils import StorageMode
+from requre.helpers.guess_object import Guess, GUESS_STR
+from requre.helpers.simple_object import Void, Simple
+from requre.online_replacing import apply_decorator_to_all_methods, replace_module_match
+from tests.testbase import BaseClass
+from tests.test_object import OwnClass
+import sys
+
+
+class Unit(BaseClass):
+    def testVoid(self):
+        self.assertEqual(Void, Guess.guess_type(sys.__stdout__))
+
+    def testSimple(self):
+        self.assertEqual(Simple, Guess.guess_type("abc"))
+        self.assertEqual(Simple, Guess.guess_type({1: 2, "a": "b"}))
+        self.assertNotEqual(ObjectStorage, Guess.guess_type({1: 2, "a": "b"}))
+
+    def testPickle(self):
+        self.assertEqual(ObjectStorage, Guess.guess_type(OwnClass(1)))
+        self.assertNotEqual(Void, Guess.guess_type(OwnClass(1)))
+        self.assertNotEqual(Simple, Guess.guess_type(OwnClass(1)))
+
+
+def obj_return(num, obj):
+    _ = num
+    return obj
+
+
+class Store(BaseClass):
+    def testFunctionDecorator(self):
+        """
+        Check if it is able to guess and store/restore proper values with types
+        """
+        decorated_own = Guess.decorator(cassette=self.cassette, item_list=[0])(
+            obj_return
+        )
+        before1 = decorated_own(1, "abc")
+        before2 = decorated_own(2, OwnClass(2))
+        before3 = decorated_own(2, OwnClass(3))
+        before4 = decorated_own(3, sys.__stdin__)
+        self.cassette.dump()
+        self.cassette.mode = StorageMode.read
+
+        after2 = decorated_own(2, OwnClass(2))
+        self.assertEqual(
+            self.cassette.data_miner.metadata[GUESS_STR], ObjectStorage.__name__
+        )
+        after3 = decorated_own(2, OwnClass(3))
+        self.assertEqual(
+            self.cassette.data_miner.metadata[GUESS_STR], ObjectStorage.__name__
+        )
+        after1 = decorated_own(1, "abc")
+        self.assertEqual(self.cassette.data_miner.metadata[GUESS_STR], Simple.__name__)
+        after4 = decorated_own(3, sys.__stdin__)
+        self.assertEqual(self.cassette.data_miner.metadata[GUESS_STR], Void.__name__)
+
+        self.assertEqual(before1, after1)
+        self.assertEqual(before2.__class__.__name__, after2.__class__.__name__)
+        self.assertEqual(before3.__class__.__name__, after3.__class__.__name__)
+        self.assertEqual(after2.__class__.__name__, "OwnClass")
+        self.assertEqual(before4.__class__.__name__, "TextIOWrapper")
+        self.assertEqual(after4.__class__.__name__, "str")
+
+
+@apply_decorator_to_all_methods(replace_module_match(what="math.sin"))
+class ApplyDefaultDecorator(unittest.TestCase):
+    SIN_OUTPUT = 0.9974949866040544
+
+    def cassette_setup(self, cassette):
+        self.assertEqual(cassette.storage_object, {})
+        # workaround to have cassette accesible inside test
+        self.cassette = cassette
```

Comment:
I am not sure if this works for the class decorator, but you can use `cassette` argument in a test like this:

https://github.com/packit/requre/blob/363c4cfcb486c5bb9ddf29bc9db2409201d73557/tests/test_online_replacing.py#L202-L207

If it does not work, we should fix it...

Reply:
Don't forget to put the type annotation -- we send it only when `cassette` argument is present with `Cassette` type annotation.

Reply:
I think that it should not work, or at least I think we do not though about this syntax, but maybe it is supported somehow :-) probably in new PR. But yep, probably it is possible to do that now. I have to try it.

Reply:
It's tested only for a single test-cases, but it should work well also for the class scenario. (We need to add a test case for that.)

Reply:
yep, it works well. this testcase also now test this ability.